Optimizely Cookie

HarveyPeachey opened the pull request:
Resolves #9734

**Overall change:**
Adds the `ckns_mvt` which is used as a user ID for optimizely experiments.

**Code changes:**

- Adds a similar implementation to [getAtUserId](https://github.com/bbc/simorgh/blob/d44f7a24f5ef750dd94ca0e3b6e9e1e738cb837a/src/app/lib/analyticsUtils/index.js#L184-L210)
- Adds a check to see if cookie has already been set, if not sets and returns the cookie value which is a UUID.
- Removes option boolean parameter `noUserId` for a streamlined approach at detecting whether the user ID should be set, if storybook running, or the live environment variable is present.

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] I have added the `cross-team` label to this PR if it requires visibility across World Service teams
- [x] I have assigned this PR to the Simorgh project
- [x] (BBC contributors only) This PR follows the [repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

**Testing:**

- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false yarn test:e2e:interactive`)
- [ ] This PR requires manual testing

HarveyPeachey committed:
adds basic cookie implementation

jroebu14 commented on src/app/containers/PageHandlers/withOptimizelyProvider/index.jsx:
```diff
@@ -19,7 +38,7 @@ const withOptimizelyProvider = (Component, noUserId = false) => {
         timeout={500}
         user={{
-          id: noUserId ? null : getAtUserId(),
+          id: noUserId ? null : getCknsMvt(),
```
Ternaries like this, where the truthy value is on the right, can be harder to read and takes a little more cognitive load to understand. It can be an easy mistake to read this like `if there is no user ID then get the cookie value`.

Making the condition clearer upfront and having the truthy part on the left of the ternary is best practice even if it forres you to add a `!` or make the ternary longer.

Also, I'm  a little confused by the variable `noUserId`. Does it mean there is no user ID? Or does it mean we should not user a user ID. If the former I think the var name should be `!hasUserId` or `!userIdExists`. Also if the former, then is this ternary correct? If no user ID then do nothing? Hard to tell because there is no PR description.

HarveyPeachey replied:
Yeah I agree, looking back at it, it's a poor choice of variable wording.

The variable is the former, meaning there is no user id present. This is used for storybook stories and for disabling the ID being used on live. I've reworded this now.

jroebu14 commented on src/app/containers/PageHandlers/withOptimizelyProvider/index.jsx:
```diff
@@ -1,7 +1,26 @@
+    Cookie.set(cookieName, uuid(), { expires, path: '/' });
+  }
+
+  return cookieValue;
```
this doesn't return the `uuid()` that is set on line 19. Is that correct?

HarveyPeachey replied:
Nope it's not correct, I have added the extra return statement to make this work

HarveyPeachey committed:
returns cookie uuid when it's being set

HarveyPeachey committed:
redefined noUserId to be less confusing

andrewscfc commented on src/app/containers/PageHandlers/withOptimizelyProvider/index.jsx:
```diff
@@ -1,15 +1,36 @@
+
+  const cookieName = 'ckns_mvt';
+  const cookieValue = Cookie.get(cookieName);
+  const expires = 365; // expires in 12 Months
```
Did you confirm this expiry with product?

HarveyPeachey replied:
The issue states it should be a year. I'll double-check with product when they're back to confirm it

pvaliani replied:
What happens in the case of a leap year? 🤔

Would we have to set a condition for this scenario?

pvaliani replied:
@JonBeeb confirmed that expiry is 12 months i.e 365 days.

andrewscfc commented on src/app/containers/PageHandlers/withOptimizelyProvider/index.jsx:
```diff
@@ -1,15 +1,36 @@
+
+  if (!cookieValue) {
+    const cookieUuid = uuid();
+    Cookie.set(cookieName, uuid(), { expires, path: '/' });
```
Does the value being a `uuid` align with other teams usages for the cookie?

HarveyPeachey replied:
Yep, I won't link to the repos, but you can search the organisation for that cookie in code and they all use the uuid v4 standard

jroebu14 commented on src/app/containers/PageHandlers/withOptimizelyProvider/index.jsx:
```diff
@@ -19,7 +40,7 @@ const withOptimizelyProvider = (Component, noUserId = false) => {
         timeout={500}
         user={{
-          id: noUserId ? null : getAtUserId(),
+          id: hasUserId && getCknsMvt(),
```
so if there is no userId then false is ok as a value here?

pvaliani replied:
Yep, if the userId is false the if statement in the ReactSDK for the provider in our case will evaluate correctly as it can handle any falsey value:

https://github.com/optimizely/react-sdk/blob/2731ec8cdfed33fae71584324babb5b371033eb4/src/Provider.tsx#L48

HarveyPeachey replied:
We've changed the implementation now, so it will in fact return null as we have removed the boolean parameter from you're suggestion

jroebu14 commented on src/app/containers/PageHandlers/withOptimizelyProvider/index.jsx:
```diff
@@ -1,15 +1,36 @@
 });
 
-const withOptimizelyProvider = (Component, noUserId = false) => {
+const withOptimizelyProvider = (Component, hasUserId = true) => {
```
another clean code best practice is to avoid [passing booleans as arguments](https://medium.com/@amlcurran/clean-code-the-curse-of-a-boolean-parameter-c237a830b7a3)

Can we not determine if there is a userId inside of this function?

pvaliani replied:
We have refactored this so that instead of setting `hasUserId` as above, we instead check to see if storybook is loaded on the page or if Optimizely is on live in order to set the userId to null - seems to be working ok on dev and storybook!

pvaliani committed:
Merge branch 'latest' into optimizely-cookie

pvaliani committed:
refactors boolean logic to ensure id is not set when storybook is on the page or when optimizely is on live

HarveyPeachey committed:
Merge branch 'latest' into optimizely-cookie

HarveyPeachey committed:
Merge branch 'latest' into optimizely-cookie

jroebu14 commented on src/app/containers/PageHandlers/withOptimizelyProvider/index.jsx:
```diff
@@ -4,31 +4,61 @@ import {
   setLogger(null);
 }
 
+const getCknsMvt = () => {
```
could this have better name? `getOpimizelyUserId` or something?

HarveyPeachey committed:
renamed optimizely cookie function to better reflect what it does

HarveyPeachey committed:
Merge branch 'optimizely-cookie' of https://github.com/bbc/simorgh into optimizely-cookie

HarveyPeachey committed:
sets cookie as defined const instead of calling uuid again

HarveyPeachey committed:
Merge branch 'latest' into optimizely-cookie

andrewscfc reviewed (commented):
Any reason there aren't any tests on this PR? `src/app/containers/PageHandlers/withOptimizelyProvider/index.jsx` seems like it should have some tests for setting the cookie. I think we have tests for similar functionality in the ATI analytics?

HarveyPeachey committed:
moved getOptimizelyUserId to own directory for ease of unit testing

HarveyPeachey committed:
re-organise withOptimizely unit tests

HarveyPeachey committed:
adds optimizely cookie unit tests

HarveyPeachey commented:
> Any reason there aren't any tests on this PR? `src/app/containers/PageHandlers/withOptimizelyProvider/index.jsx` seems like it should have some tests for setting the cookie. I think we have tests for similar functionality in the ATI analytics?

Nope, forgot to add some. I've now added unit tests, also extracted it out into it's own export for ease of testing and to keep it a bit more organised

HarveyPeachey committed:
Merge branch 'latest' into optimizely-cookie

andrewscfc reviewed (approved):
Thanks for the tests @HarveyPeachey

HarveyPeachey committed:
Merge branch 'latest' into optimizely-cookie

paruchurisilpa committed:
Merge branch 'latest' into optimizely-cookie

paruchurisilpa commented:
Checked using the link - http://localhost:7080/arabic/23298105

I can see ckns_mvt cookie which expires in 1 year..

<img width="1434" alt="Screenshot 2022-01-19 at 13 09 40" src="https://user-images.githubusercontent.com/1379277/150137377-dc55c7ea-3d34-4282-aea2-84764dab64a1.png">

Once I delete the cookie and reload the page,  ckns_mvt cookie is created with a new id is..

<img width="1433" alt="Screenshot 2022-01-19 at 13 13 41" src="https://user-images.githubusercontent.com/1379277/150137980-d46c636b-fb49-4823-a893-e53e26f2f757.png">

It shouldn't set a cookie in Storybook. And on the live, it won’t set the cookie. This scenario  is covered by unit test.

Rest LGTM...

HarveyPeachey committed:
Merge branch 'latest' into optimizely-cookie

HarveyPeachey committed:
Merge branch 'latest' of https://github.com/bbc/simorgh into optimizely-cookie

HarveyPeachey committed:
updates bundle size

HarveyPeachey committed:
Merge branch 'optimizely-cookie' of https://github.com/bbc/simorgh into optimizely-cookie
